src/main/java/app/linguistai/bmvp/service/gamification/quest/QuestService.java

```diff
@@ -56,8 +77,162 @@ public Boolean checkUserHasQuestType(String email, QuestCompletionCriteria type)
+
+            // Step 2. Check if user has active quests (assigned that day)
+            if (this.userHasActiveQuests(user)) {
+                return; // Step 2.1 If user has active quests for today, return without modification
+            }
+
+            // Step 2.2 If the user does not have any quests/quests are older than a day, remove the old quests
+            this.deleteInactiveQuests(user);
+
+            // Step 3. Assign 2 "Word Practice" quests for different words
+            List<Quest> useWordQuests = this.buildBulkUseWordQuests(user, 2);
+
+            // Step 4. Assign a "Send Message" quest
+            Integer randomSendMessagedTimes = this.generateRandomTimesForUseWordQuest();
+            Quest sendMessage = this.buildSendMessageQuest(user, randomSendMessagedTimes);
+
+            // Step 5. Save all quests to the repository
+            List<Quest> questsToAssign = new ArrayList<>(useWordQuests);
+            questsToAssign.add(sendMessage);
+
+            questRepository.saveAll(questsToAssign);
+        }
+        catch (Exception e) {
+            System.out.println("ERROR: Could not assign quests to user with email " + email + ".");
+            throw e;
+        }
+    }
+
+    private List<Quest> buildBulkUseWordQuests(User user, Integer count) throws Exception {
+        if (count <= 0) {
+            throw new IllegalArgumentException("ERROR: Build bulk use word quests method failed. Count must be greater than zero.");
+        }
+
+        List<Quest> useWordQuests = new ArrayList<>();
+        HashSet<String> usedWords = new HashSet<>();
+
+        for (int i = 0; i < count; i++) {
+            // Get one active unknown word list
+            UnknownWordList randomActiveList = unknownWordService.getRandomActiveUnknownWordList(user.getId());
+
+            // Select a random word from the retrieved unknown word list, while ensuring no duplicates exist for quests
+            String randomWord = this.returnUniqueRandomWordForUseWordQuest(randomActiveList.getListId(), usedWords);
+            usedWords.add(randomWord);
+
+            // Generate a random "times" field
+            Integer randomUseWordTimes = this.generateRandomTimesForUseWordQuest();
+
+            // Build the Use Word quest and add to the results list
+            useWordQuests.add(this.buildUseWordQuest(user, randomWord, randomUseWordTimes));
+        }
+
+        return useWordQuests;
+    }
+
+    /**
+     * Select a random word from the retrieved unknown word list, while ensuring no duplicates exist for quests
+     * @param listId List ID for which the words will be pulled from
+     * @param usedWordsSoFar Used words so far in the generation
+     * @return Random unique word for "Use Word" type quests
+     * @throws Exception When a random word cannot be retrieved from list with ID listId
+     */
+    private String returnUniqueRandomWordForUseWordQuest(UUID listId, HashSet<String> usedWordsSoFar) throws Exception {
```

Comment:
I feel like something such as 
```
Collections.shuffle(items);
List<String> randomItems = items.subList(0, 2);
```
would be more efficient and more readable. But this works fine too, your call

Reply:
Thanks for the recommendation, overloaded a function in the `UnknownWord` repository to handle this with: "`getRandomWordFromList(UUID listId, Integer numOfWords)`"